Mathlib/LinearAlgebra/TensorProduct/Tower.lean

```diff
@@ -176,27 +183,210 @@ nonrec def mk : M →ₗ[A] N →ₗ[R] M ⊗[R] N :=
 #align tensor_product.algebra_tensor_module.mk TensorProduct.AlgebraTensorModule.mk
 #align tensor_product.algebra_tensor_module.mk_apply TensorProduct.AlgebraTensorModule.mk_apply
 
+variable {R A B M N P Q}
+
+/-- Heterobasic version of `TensorProduct.map` -/
+def map (f : M →ₗ[A] P) (g : N →ₗ[R] Q) : M ⊗[R] N →ₗ[A] P ⊗[R] Q :=
+  lift $ (show (Q →ₗ[R] P ⊗ Q) →ₗ[A] N →ₗ[R] P ⊗[R] Q from
+  { toFun := fun h => h ∘ₗ g,
+    map_add' := fun h₁ h₂ => LinearMap.add_comp g h₂ h₁,
+    map_smul' := fun c h => LinearMap.smul_comp c h g }) ∘ₗ mk R A P Q ∘ₗ f
+
+@[simp] theorem map_tmul (f : M →ₗ[A] P) (g : N →ₗ[R] Q) (m : M) (n : N) :
+    map f g (m ⊗ₜ n) = f m ⊗ₜ g n :=
+  rfl
+
+theorem map_add_left (f₁ f₂ : M →ₗ[A] P) (g : N →ₗ[R] Q) :
+    map (f₁ + f₂) g = map f₁ g + map f₂ g := by
+  ext
+  simp_rw [curry_apply, TensorProduct.curry_apply, restrictScalars_apply, add_apply, map_tmul,
+    add_apply, add_tmul]
+
+theorem map_add_right (f : M →ₗ[A] P) (g₁ g₂ : N →ₗ[R] Q) :
+    map f (g₁ + g₂) = map f g₁ + map f g₂ := by
+  ext
+  simp_rw [curry_apply, TensorProduct.curry_apply, restrictScalars_apply, add_apply, map_tmul,
+    add_apply, tmul_add]
+
+theorem map_smul_right (r : R) (f : M →ₗ[A] P) (g : N →ₗ[R] Q) : map f (r • g) = r • map f g := by
+  ext
+  simp_rw [curry_apply, TensorProduct.curry_apply, restrictScalars_apply, smul_apply, map_tmul,
+    smul_apply, tmul_smul]
+
+theorem map_smul_left (r : B) (f : M →ₗ[A] P) (g : N →ₗ[R] Q) : map (r • f) g = r • map f g := by
+  ext
+  simp_rw [curry_apply, TensorProduct.curry_apply, restrictScalars_apply, smul_apply, map_tmul,
+    smul_apply, smul_tmul']
+
+variable (R A B M N P Q)
+
+/-- Heterobasic version of `TensorProduct.map_bilinear` -/
+def mapBilinear : (M →ₗ[A] P) →ₗ[B] (N →ₗ[R] Q) →ₗ[R] (M ⊗[R] N →ₗ[A] P ⊗[R] Q) :=
+  { toFun := fun f =>
+    { toFun := fun g => map f g
+      map_add' := fun _g₁ _g₂ => map_add_right _ _ _
```

Comment:
```suggestion
      map_add' := fun _g₁ _g₂ => map_add_right ..
```
I've only just learnt about this notation today or yesterday on Zulip so I might be a bit overenthusiastic about it, but don't you think these next four lines are as good a place as any to use it? Or do you think it's confusing?